### ISSUE TITLE
An attempt to download playlists results in an accumulation of long list in Tasks " unavailable videos" from all current and previousl attempts.y failed downloads 

### ISSUE THREAD

**Issue by EMG70** [iiab/calibre-web#238]
## Short Notice from the maintainer 

After 6 years of more or less intensive programming on Calibre-Web, I need a break. 
The last few months, maintaining Calibre-Web has felt more like work than a hobby. I felt pressured and teased by people to solve "their" problems and merge PRs for "their" Calibre-Web. 
I have turned off all notifications from Github/Discord and will now concentrate undisturbed on the development of “my” Calibre-Web over the next few weeks/months.  
I will look into the issues and maybe also the PRs from time to time, but don't expect a quick response from me.

Please also have a look at our [Contributing Guidelines](https://github.com/janeczku/calibre-web/blob/master/CONTRIBUTING.md)

**Describe the bug/problem** 
An attempt to download playlists results in an accumulation of long list in Tasks " unavailable videos" from all current and previously attempts. 

A clear and concise description of what the bug is. If you are asking for support, please check our [Wiki](https://github.com/janeczku/calibre-web/wiki) if your question is already answered there.

Unavailable videos for any reason are expected to show as those that have failed for the playlist not cumulative list from previous downloads.
**To Reproduce**

Download 
1. Go to 'https://www.youtube.com/@agribusinesswithwadzanaima275' name of playlist
2. Click on 'https://www.youtube.com/watch?v=EiMFnDquvdM' failing video being added to the list of unavailable videos
3. https://www.youtube.com/watch?v=b4RLo3tgpoA failing video being added to the list of unavailable videos
4. Scroll down to '....'
5. See error

**Logfile**
SUDO IIAB-DIAGNOSTICS - https://paste.centos.org/view/10fc9a97

[calibre-web.log](https://github.com/user-attachments/files/16679173/calibre-web.log)

Add content of calibre-web.log file or the relevant error, try to reproduce your problem with "debug" log-level to get more output.

**Expected behavior**

Unavailable videos for any reason are expected to show as those that have failed for the playlist not cumulative list from previous downloads.

**Screenshots**
![Screenshot from 2024-08-20 17-34-33](https://github.com/user-attachments/assets/ef30fc76-016a-418b-8b9b-a742d9d3df15)

**Environment (please complete the following information):**

 - OS: [e.g. Windows 10/Raspberry Pi OS]
 - Python version: [e.g. python2.7]
 - Calibre-Web version: [e.g. 0.6.8 or 087c4c59 (git rev-parse --short HEAD)]:
 - Docker container: [None/LinuxServer]:
 - Special Hardware: [e.g. Rasperry Pi Zero]
 - Browser: [e.g. Chrome 83.0.4103.97, Safari 13.3.7, Firefox 68.0.1 ESR]
![Screenshot from 2024-08-20 17-37-02](https://github.com/user-attachments/assets/89b65faa-da98-4f1a-96bb-c5ac3aa8d22f)

**Additional context**
Add any other context about the problem here. [e.g. access via reverse proxy, database background sync, special database location]
This is a members only video https://www.youtube.com/watch?v=MfRdmsbs3R8 but also gets added to list of unavailable accumulative list.
![Screenshot from 2024-08-20 17-45-35](https://github.com/user-attachments/assets/58ad764d-385e-463c-bccd-af399d340d2d)

This video https://www.youtube.com/watch?v=EiMFnDquvdM fails to download as part of playlist https://www.youtube.com/@agribusinesswithwadzanaima275 but successfully downloaded as an individual url.(last line of screenshot)
![Screenshot from 2024-08-20 17-40-12](https://github.com/user-attachments/assets/7cbdeffb-e4e0-4002-a666-ff7fcfb0b237)

**Comment by deldesir** [comment]
Here is why:
https://github.com/iiab/calibre-web/blob/486b7bbcb5b5450c81850846f771f329d3dc238b/scripts/lb-wrapper#L82
This line of code from lb-wrapper ensures, with the use of `--force`, to update already existed metadata for a given path as explained in https://github.com/chapmanjacobd/library/blob/main/xklb/usage.py#L570-L572
- When a video url is submitted, the path is the video url, the metadata for this url is updated. The download goes through
- When a Youtube Playlist is submitted, the path is the playlist url, the metadate for the playlist url is updated, but the video urls path are not because they are secondary. Previous failures are seen and handled then.

**Comment by holta** [comment]
@deldesir which below are most relevant?  (In helping us explain — and decide what to do next!)

- PR #125 (added `--force` to tubeadd)
- #150
  - PR #151

**Comment by holta** [comment]
@EMG70 just FYI @deldesir is looking into this, but it might not be easy!

Something equivalent to xklb tubeadd's `--force` might be needed for entire playlists — not just single videos ?

(Sorting how xklb helps us with download errors & metadata-extract.py for entire playlists — not just single videos — likely needs work!)

**Comment by EMG70** [comment]
> @EMG70 just FYI @deldesir is looking into this, but it might not be easy!
> 
> Something equivalent to xklb tubeadd's `--force` might be needed for entire playlists — not just single videos ?
> 
> (Sorting how xklb helps us with download errors & [metadata-extract.py](https://github.com/iiab/calibre-web/blob/486b7bbcb5b5450c81850846f771f329d3dc238b/cps/tasks/metadata_extract.py#L70) for entire playlists — not just single videos — likely needs work!)

Thank you

**Comment by EMG70** [comment]
Yes I was aware of her 220 agricultural videos under her channel ,unfortunately she doesn't group them by Playlist.I have downloaded same videos successfully before my calibre-web got corrupted after upgrading to 24.04 in last two weeks .
It could well be the new Youtube restrictions which came up recently.

**Comment by holta** [comment]
@deldesir suspects the large number of videos (220 in this case) might be causing this problem?

Regardless whether or not contained in a standard playlist.

**Comment by deldesir** [comment]
Yes, the number of videos might trigger a restriction. 220 videos were processed to get their metadata extracted so 100 of them could be downloaded. Tests with single videos downloaded in another VM in another location went through without a problem.

**Comment by EMG70** [comment]
Thank you ,I think that explains it .This said calibre-web /server  was built circa March 2024, probably before the 100 video limit was implemented ,so explains why i initially managed. I guess things changed when i ran sudo iiab-update-f.

**Comment by holta** [comment]
@EMG70 can you please retest, on a completely clean/fresh VM?  @deldesir should investigate & test wherever he can help shed light, as I just tried "Download to IIAB" with https://www.youtube.com/@agribusinesswithwadzanaima275 and the 220 videos are downloading perfectly so far... (100% success with the first 10 videos anyway!)

![image](https://github.com/user-attachments/assets/fc450f41-9140-47b7-88e0-4eaa91dd465e)

**Comment by EMG70** [comment]
Will do

**Comment by holta** [comment]
> Will do

Thanks!  FYI @EMG70 33 videos (of 220) downloaded successfully over about 30min, so I cut it off by running `systemctl restart calibre-web`

PS I could have run `sudo rm /library/calibre-web/xklb-metadata.db` to force the remaining ~187 videos no to try again — which is likely to happen the next time "Download to IIAB" is attempted!

**Comment by EMG70** [comment]
Tested with a new VM ,100 videos successfully downloaded from the channel https://www.youtube.com/@agribusinesswithwadzanaima275.
![Screenshot from 2024-08-29 12-26-18](https://github.com/user-attachments/assets/d4b59d0d-f005-47ec-bf62-295c2a5858ae)

[calibre-web.log](https://github.com/user-attachments/files/16798280/calibre-web.log)

**Comment by holta** [comment]
@deldesir should (speculate!) as to the true root cause:

What really changed since a week ago to cause this to start working again?

**Comment by holta** [comment]
@EMG70 when exactly did your try downloading 300 instead of 100 videos?

(Thanks for clarifying, just in case this happens s related!)

**Comment by holta** [comment]
> @EMG70 when exactly did your try downloading "up to 300 videos" instead of "up to 100 videos" ?

@EMG70 is the timeline/summary below accurate, to help us all understand?

- Prior to 2024-08-06, @EMG70 tried downloading "up to 300 videos" on an Ubuntu 23.10 machine, that on 2024-08-06 had problems upgrading to Ubuntu 24.04.
  - _(While very unfortunate indeed... that would appear to be... completely separate from understanding this issue #238 here ??)_

- On 2024-08-19 and 2024-08-20, @EMG70 tried and failed to download of [youtube.com/@agribusinesswithwadzanaima275](https://youtube.com/@agribusinesswithwadzanaima275) to his **newly installed IIAB** on Ubuntu Desktop 24.04 (i3-4130T, 8G RAM, 469G disk): https://paste.centos.org/view/10fc9a97

**Comment by holta** [comment]
@deldesir please help us understand what could have caused most of 100 videos to fail to download (`Unavailable Video(s):`) on 2024-08-20 (this issue `#238` !)

- _Speculative ideas need to be circulated, to promote deeper understanding among all, in case this-or-similar happens again!_

- ASIDE: @EMG70 suspects that if he had not updated calibre-web on his old 23.10... he would still be able to download above 100 videos?  (Which certainly interesting, and somewhat tragic that he lost a few "bookshelves" of downloaded videos, that's appears to be 100% off-topic to this specific issue right here!)

**Comment by holta** [comment]
@EMG70 asks:

**Comment by holta** [comment]
@deldesir please help @EMG70 & all of us deepen our intuition as to what might-or-might-not have been happening to cause this last week!

@EMG70 asks:

> This accumulation of failed videos [ `Unavailable Video(s):` ]  was also happening on a VM last week when i tried same channel,probably meaning something changed on latest script ?

**Comment by deldesir** [comment]
> @deldesir please help us understand what could have caused most of 100 videos to fail to download (`Unavailable Video(s):`) on 2024-08-20 (this issue `#238` !)

Since reproducibility of the issue is not possible anymore and the log doesn't contain an error relevant to the failures, I am afraid we'll have to rely on untested assumptions.

What I know is at the time of the bug, yt-dlp 2024.08.06 and xklb 2.9.026 were in use.

**Comment by holta** [comment]
> at the time of the bug, yt-dlp 2024.08.06 and xklb 2.9.026 were in use.

- yt-dlp 2024.08.06 ✅
- xklb 2.9.026 ⚠️

Ok!  xklb 2.9.026 could easily be tested to check if that might be the cause?

As outlined in https://github.com/iiab/calibre-web/wiki#xklb-and-yt-dlp-backend-integration — everyone should do this regularly to keep assumptions grounded in reality:

> To revert to a prior version, e.g. 2.9.026:
>
> `pipx install --force xklb==2.9.026`

**Comment by holta** [comment]
> To revert to a prior version, e.g. 2.9.026:
> 
> `pipx install --force xklb==2.9.026`

_Much safer method below!_ (as [wiki doc](https://github.com/iiab/calibre-web/wiki#xklb-and-yt-dlp-backend-integration) explains)

```
pipx uninstall xklb
pipx install xklb==2.9.026
```

**Comment by deldesir** [comment]
Tested with xklb 2.9.026 without a problem. Might be related to xklb-metadata.db being corrupted at the time.
![image](https://github.com/user-attachments/assets/d91743a9-2140-439d-98e5-562aafdde519)

**Comment by EMG70** [comment]
Thanks all .Am i right in saying one can edit code in /usr/local/calibre-web-py3/cps/constants.py if they really need to download more than 100 videos ?
as on screenshot below:
![IMG_4833](https://github.com/user-attachments/assets/30153020-4b4a-4350-8fe0-8c86509eee3d)

**Comment by holta** [comment]
> /usr/local/calibre-web-py3/cps/constants.py

Yes this is off-topic :smile:

But of course feel free to try!  And if YouTube eventually blocks your IP address (e.g. #243 or similar), or you run into any other problems, definitely let us know!

RECAP: @EMG70 modified `MAX_VIDEOS_PER_DOWNLOAD = 100` to `MAX_VIDEOS_PER_DOWNLOAD = 300` in the past... and everyone is free to try doing similar...

https://github.com/iiab/calibre-web/blob/486b7bbcb5b5450c81850846f771f329d3dc238b/cps/constants.py#L51-L52
